### PR TITLE
Add Local Optima Metric

### PR DESCRIPTION
**Pull request by Stargazer1998**
create local optima and its helper functions

## Describe your changes
Local Optima metric and its helper functions added
## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] If it is a core feature, I have added thorough tests.

### CODE REVIEW

File: src/tum_dlr_automl_for_eo/utils/custom_metrics.py [hunk 1]
```diff
@@ -1,4 +1,102 @@
...
+    @param architectures: np.array [(Number of Models) x (length of binary encoded architecture)] binary encoded
+    architectures
+    @param acc_list: np.array [Number of Models] Final Accuracies of Models @param starting_point: int
+    Starting model to search
+    @return: list [found architecture index, found architecture's accuracy]
+             in case of no improvement returns given architecture info.
+    """
+    MAX_NUMBER_OF_ITERATIONS = 50
+
+    architectures = architectures
```
Comment (emreds) [inline src/tum_dlr_automl_for_eo/utils/custom_metrics.py:37]: Is it necessary to define like this?

File: src/tum_dlr_automl_for_eo/utils/custom_metrics.py [hunk 2]
```diff
@@ -1,4 +1,102 @@
...
+    """
+
+    @param architectures: np.array [(Number of Models) x (length of binary encoded architecture)] binary encoded
+    architectures
+    @param acc_list: np.array [Number of Models] Final Accuracies of Models @param starting_point: int
+    Starting model to search
+    @return: list [found architecture index, found architecture's accuracy]
+             in case of no improvement returns given architecture info.
+    """
+    MAX_NUMBER_OF_ITERATIONS = 50
```
Comment (emreds) [inline src/tum_dlr_automl_for_eo/utils/custom_metrics.py:35]: What do you think about making this a parameter?

**Comment by JahodaPaul** [comment]
search_local_optima function should take in the number of neighbors and MAX_Iteration as a parameter. The function should also return the final number of local optima estimation.. where is the BILS algorithm estimation? Otherwise great!

File: src/tum_dlr_automl_for_eo/utils/custom_metrics.py (hunk 2 above)
Comment (Stargazer1998) [reply src/tum_dlr_automl_for_eo/utils/custom_metrics.py:35]: i think this is something to be decided as a result of cost analysis and will be a fixed parameter, but ideally we can

File: src/tum_dlr_automl_for_eo/utils/custom_metrics.py [hunk 3]
```diff
@@ -1,4 +1,102 @@
...
+
+        # Loop over obtained 5 neighbors
+        for neighbor in neighbors:
+            for arc_index, arc in enumerate(architectures):
+
+                if all(neighbor == arc) and acc_list[arc_index] > current_acc:
+                    found_arc_index = arc_index
+                    found_acc = acc_list[found_arc_index]
+                    # print(
+                    #     f"from:{curr_architecture_index} to:{found_arc_index} pre_acc:{current_acc} found_acc:{found_acc}")
```
Comment (emreds) [inline src/tum_dlr_automl_for_eo/utils/custom_metrics.py:59]: I just wanted to ask the understand the BILS algorithm. In here, we compare the neighbours of an architecture with the all previous architectures and if the existing architecture has better accuracy we count that as a found architecture right?

Comment (Stargazer1998) [reply src/tum_dlr_automl_for_eo/utils/custom_metrics.py:59]: Yes exactly

File: src/tum_dlr_automl_for_eo/utils/custom_metrics.py (hunk 2 above)
Comment (Stargazer1998) [reply src/tum_dlr_automl_for_eo/utils/custom_metrics.py:35]: This is something I will modify

**Review by emreds — APPROVED** [review]

**Pull request merged**